src/dragonball/src/device_manager/mod.rs

```
@@ -0,0 +1,600 @@
+    }
+}
+
+/// Context for device addition/removal operations.
+pub struct DeviceOpContext {
+    epoll_mgr: Option<EpollManager>,
+    io_context: DeviceManagerContext,
+    irq_manager: Arc<KvmIrqManager>,
+    res_manager: Arc<ResourceManager>,
+    vm_fd: Arc<VmFd>,
+    vm_as: Option<GuestAddressSpaceImpl>,
+    address_space: Option<AddressSpace>,
+    logger: slog::Logger,
+    is_hotplug: bool,
+
+    #[cfg(feature = "dbs-virtio-devices")]
+    virtio_devices: Vec<Arc<DbsMmioV2Device>>,
+}
+
+impl DeviceOpContext {
+    pub(crate) fn new(
+        epoll_mgr: Option<EpollManager>,
+        device_mgr: &DeviceManager,
+        vm_as: Option<GuestAddressSpaceImpl>,
+        address_space: Option<AddressSpace>,
+        is_hotplug: bool,
+    ) -> Self {
+        let irq_manager = device_mgr.irq_manager.clone();
+        let res_manager = device_mgr.res_manager.clone();
+
+        let vm_fd = device_mgr.vm_fd.clone();
+        let io_context = DeviceManagerContext {
+            io_manager: device_mgr.io_manager.clone(),
+            io_lock: device_mgr.io_lock.clone(),
+        };
+        let logger = device_mgr.logger.new(slog::o!());
+
+        DeviceOpContext {
+            epoll_mgr,
+            io_context,
+            irq_manager,
+            res_manager,
+            vm_fd,
+            vm_as,
+            address_space,
+            logger,
+            is_hotplug,
+            #[cfg(feature = "dbs-virtio-devices")]
+            virtio_devices: Vec::new(),
+        }
+    }
+
+    pub(crate) fn get_vm_as(&self) -> Result<GuestAddressSpaceImpl> {
+        match self.vm_as.as_ref() {
+            Some(v) => Ok(v.clone()),
+            None => Err(DeviceMgrError::InvalidOperation),
+        }
+    }
+
+    pub(crate) fn logger(&self) -> &slog::Logger {
+        &self.logger
+    }
+
+    fn generate_kernel_boot_args(&mut self, kernel_config: &mut KernelConfigInfo) -> Result<()> {
+        assert!(!self.is_hotplug);
```

This looks potentially dangerous for production code. Could you return an error instead and add a unit test to assert this error path?

---

Changed to return an error.
We'll introduce unit tests for the device manager part in later PRs so I'll add a unit test for this function in future PRs. #4410 issue is created for this.